wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceService.java

```diff
@@ -656,6 +667,41 @@ public void deleteProject(@ApiParam("The workspace id")
         }
     }
 
+    @GET
+    @Path("/{id}/check")
+    @Produces(APPLICATION_JSON)
+    @ApiOperation(value = "Get state of the workspace agent by the workspace id and agent id")
+    @ApiResponses({@ApiResponse(code = 200, message = "The response contains requested workspace entity"),
+                   @ApiResponse(code = 400, message = "Missed required parameters, parameters are not valid"),
+                   @ApiResponse(code = 403, message = "The user is not workspace owner"),
+                   @ApiResponse(code = 404, message = "The workspace with specified id does not exist"),
+                   @ApiResponse(code = 500, message = "Internal server error occurred")})
+    public WsAgentHealthStateDto checkAgentHealth(@ApiParam(value = "Workspace id")
+                                                  @PathParam("id") String key) throws NotFoundException,
+                                                                                             ServerException,
+                                                                                             ForbiddenException,
+                                                                                             BadRequestException,
+                                                                                             UnauthorizedException,
+                                                                                             IOException,
+                                                                                             ConflictException {
+        final WorkspaceImpl workspace = workspaceManager.getWorkspace(key);
+        if (WorkspaceStatus.RUNNING != workspace.getStatus()) {
+            return newDto(WsAgentHealthStateDto.class).withWorkspaceStatus(workspace.getStatus());
+        }
+
+        final MachineImpl devMachine = workspace.getRuntime().getDevMachine();
+        if (devMachine == null) {
+            return newDto(WsAgentHealthStateDto.class)
+                    .withWorkspaceStatus(workspace.getStatus())
+                    .withCode(NOT_FOUND.getStatusCode())
+                    .withReason("Workspace Agent isn't available if Dev machine isn't RUNNING");
```

Comment:
I think the most modern term ws-machine, but I'm not sure. @eivantsov can you help with that?